nntrainer/src/concat_layer.cpp

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file	concat_layer.cpp
+ * @date	27 Oct 2020
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug		No known bugs except for NYI items
+ * @brief	This is Concat Layer Class for Neural Network
+ *
+ */
+
+#include <concat_layer.h>
+#include <cstring>
+#include <layer_internal.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+int ConcatLayer::initialize() {
+  int status = ML_ERROR_NONE;
+  unsigned int channel = 0;
+  if (num_inputs == 0) {
+    ml_loge("Error: number of inputs are not initialized");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    channel += input_dim[idx].channel();
+  }
+
+  output_dim[0] = input_dim[0];
+  output_dim[0].channel(channel);
+
+  return status;
+}
+
+sharedConstTensors ConcatLayer::forwarding(sharedConstTensors in) {
+
+  TensorDim d = in[0]->getDim();
+  unsigned int concat_channel = d.channel();
+
+  for (unsigned int idx = 1; idx < num_inputs; ++idx) {
+    TensorDim dim = in[idx]->getDim();
+
+    for (unsigned int i = 2; i < d.rank(); ++i) {
+      if (d[i] != dim[i])
```

Comment:
how about moving these rank dependent checks to initialise with input_dim and rather just compare in[i].getDim() with input_dim[i]? lets keep forwarding (critical path) least checks.

Reply:
Further, I think that these checks of the dimension of `in` and `input_dim` can be all skipped from `forwarding` as nntrainer layers passes the data to this layer. If input to the model is correct dimension, this function will automatically receive the correct dimensions. This is the reason no other layer checks the dimension of the input in forwarding/backwarding.

Reply:
Maybe it could live inside `#ifdef DEBUG`

Reply:
fixed. Thanks

Reply:
@jijoongmoon I think this check is still needed at the initialize phase to ensure that the connected layers set by the model are compatible.